### PR TITLE
Implement ConditionallySpeculatable for {Dynamic,}BroadcastInDim

### PR DESCRIPTION
**Pull request by mlevesquedion**

### CODE REVIEW

File: stablehlo/dialect/StablehloOps.cpp [hunk 1]
```diff
@@ -1343,6 +1343,27 @@ LogicalResult DynamicBroadcastInDimOp::reifyReturnTypeShapes(
   return success();
 }
 
+mlir::Speculation::Speculatability
+DynamicBroadcastInDimOp::getSpeculatability() {
+  auto operandType = getOperand().getType();
+
+  // If input is dynamic, not speculatable.
+  if (!operandType.hasStaticShape()) return mlir::Speculation::NotSpeculatable;
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.cpp:1351]: How about:
```
If input is dynamic, the broadcasting rules might be violated at runtime and hence not speculatable. 
```

File: stablehlo/dialect/StablehloOps.cpp [hunk 2]
```diff
@@ -1343,6 +1343,27 @@ LogicalResult DynamicBroadcastInDimOp::reifyReturnTypeShapes(
...
+
+  // If input is broadcastable (all 1's) and result is dynamic, speculatable.
+  auto resultDynamic =
```
Comment (sdasgup3) [inline stablehlo/dialect/StablehloOps.cpp:1354]: Can't we use something like
```
auto resultDynamic = !getType().hasSaticShape();
```

File: stablehlo/dialect/StablehloOps.cpp (hunk 1 above)
Comment (mlevesquedion) [reply stablehlo/dialect/StablehloOps.cpp:1351]: Thanks for the suggestion! Done.

**Review by sdasgup3 — APPROVED** [review]

File: stablehlo/dialect/TypeInference.cpp [hunk 3]
```diff
@@ -3714,6 +3714,23 @@ LogicalResult verifyDynamicBroadcastInDimOp(
                                " does not refer to a "
                                "valid operand dimension");
 
+  if (SmallVector<int64_t> shape;
+      operandType.hasStaticShape() &&
+      matchInts(outputDimensions, shape).succeeded()) {
+    for (int64_t i = 0; i != bcastDimensionsSize; ++i) {
```
Comment (ghpvnist) [inline stablehlo/dialect/TypeInference.cpp:3720]: I recently found out, but there's a `llvm::enumerate` to replace this.

```
for (auto [i, dimIndex] : llvm::enumerate(broadcastDimensions)
...
```

**Review by ghpvnist — APPROVED** [review]
LGTM with one comment.

File: stablehlo/dialect/TypeInference.cpp (hunk 3 above)
Comment (mlevesquedion) [reply stablehlo/dialect/TypeInference.cpp]: Good catch, thanks!

**Pull request merged**